Online signing with sigstore failed in a new repository

I created a new repository (https://github.com/jonnystoten/tuf-testing) by cloning the template, then followed the instructions in [REPOSITORY-MAINTENANCE.md](https://github.com/theupdateframework/tuf-on-ci/blob/main/docs/REPOSITORY-MAINTENANCE.md), setting up sigstore as the online signing method. The init signing event went fine but after merging the online snapshot/timestamp sign did not work.

The error was:

```
Run tuf-on-ci-online-sign --push
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/tuf/api/metadata.py", line 386, in sign
    signature = signer.sign(bytes_data)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/securesystemslib/signer/_sigstore_signer.py", line 258, in sign
    with context.signer(self._token) as sigstore_signer:
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/sigstore/sign.py", line 308, in signer
    yield Signer(identity_token, self, cache)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/sigstore/sign.py", line 124, in __init__
    self.__cached_signing_certificate = self._signing_cert(self._private_key)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/sigstore/sign.py", line 157, in _signing_cert
    NameOID.EMAIL_ADDRESS, self._identity_token._identity
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute '_identity'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.6/x64/bin/tuf-on-ci-online-sign", line 8, in <module>
    sys.exit(online_sign())
             ^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/tuf_on_ci/online_sign.py", line 51, in online_sign
    snapshot_updated, _ = repo.do_snapshot(not valid_snapshot)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/tuf/repository/_repository.py", line 191, in do_snapshot
    with self.edit_snapshot() as snapshot:
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/tuf/repository/_repository.py", line 125, in edit_snapshot
    with self.edit(Snapshot.type) as snapshot:
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/tuf/repository/_repository.py", line 104, in edit
    self.close(role, md)
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/tuf_on_ci/_repository.py", line 213, in close
    md.sign(signer, True)
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/site-packages/tuf/api/metadata.py", line 388, in sign
    raise UnsignedMetadataError("Problem signing the metadata") from e
tuf.api.exceptions.UnsignedMetadataError: Problem signing the metadata
Error: Process completed with exit code 1.
```

Looks like the [sigstore Signer](https://github.com/sigstore/sigstore-python/blob/9ca834443c05ea3660a6fc9a7759ae6866eb5356/sigstore/sign.py#L90) is being created somewhere with a string as the identity_token, perhaps just missing the call to the IdentityToken constructor? I'm not sure where this is actually happening though!

Full details here: https://github.com/jonnystoten/tuf-testing/actions/runs/6721057085/job/18266712416.

--- a/repo/tuf_on_ci/_repository.py
+++ b/repo/tuf_on_ci/_repository.py
@@ -10,12 +10,12 @@ from glob import glob
 from securesystemslib.exceptions import UnverifiedSignatureError
 from securesystemslib.signer import (
     KEY_FOR_TYPE_AND_SCHEME,
+    SIGNER_FOR_URI_SCHEME,
     Signature,
     Signer,
     SigstoreKey,
     SigstoreSigner,
 )
-from sigstore.oidc import detect_credential
 from tuf.api.exceptions import UnsignedMetadataError
 from tuf.api.metadata import (
     Key,
@@ -32,6 +32,7 @@ from tuf.repository import AbortEdit, Repository
 
 # sigstore is not a supported key by default
 KEY_FOR_TYPE_AND_SCHEME[("sigstore-oidc", "Fulcio")] = SigstoreKey
+SIGNER_FOR_URI_SCHEME[SigstoreSigner.SCHEME] = SigstoreSigner
 
 # TODO Add a metadata cache so we don't constantly open files
 # TODO; Signing status probably should include an error message when valid=False
@@ -205,11 +206,7 @@ class CIRepository(Repository):
         for key in self._get_keys(rolename):
             if rolename in ["timestamp", "snapshot"]:
                 uri = key.unrecognized_fields["x-tuf-on-ci-online-uri"]
-                # WORKAROUND while sigstoresigner is not finished
-                if uri == "sigstore:":
-                    signer = SigstoreSigner(detect_credential(), key)
-                else:
-                    signer = Signer.from_priv_key_uri(uri, key)
+                signer = Signer.from_priv_key_uri(uri, key)
                 md.sign(signer, True)
             else:
                 # offline signer, add empty sig

